pilopt/src/lib.rs

```diff
@@ -0,0 +1,416 @@
+                *e = Expression::Number(*value);
+            }
+        }
+        ControlFlow::Continue::<()>(())
+    });
+
+    pil_file.remove_polynomials(&constant_polys.keys().cloned().collect());
+}
+
+fn constrained_to_constant<T: FieldElement>(expr: &Expression<T>) -> Option<(PolyID, T)> {
+    match expr {
+        Expression::BinaryOperation(left, BinaryOperator::Sub, right) => {
+            match (left.as_ref(), right.as_ref()) {
+                (Expression::Number(n), Expression::PolynomialReference(poly))
+                | (Expression::PolynomialReference(poly), Expression::Number(n)) => {
+                    if poly.is_witness() {
+                        // This also works if "next" is true.
+                        return Some((poly.poly_id.unwrap(), *n));
+                    }
+                }
+                _ => {}
+            }
+        }
+        Expression::PolynomialReference(poly) => {
+            if poly.is_witness() {
+                return Some((poly.poly_id.unwrap(), 0.into()));
+            }
+        }
+        _ => {}
+    }
+    None
+}
+
+/// Removes identities that evaluate to zero and lookups with empty columns.
+fn remove_trivial_identities<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+    let to_remove = pil_file
+        .identities
+        .iter()
+        .enumerate()
+        .filter_map(|(index, identity)| match identity.kind {
+            IdentityKind::Polynomial => {
+                if let Expression::Number(n) = identity.left.selector.as_ref().unwrap() {
+                    if *n == 0.into() {
+                        return Some(index);
+                    }
+                    // Otherwise the constraint is not satisfiable,
+                    // but better to get the error elsewhere.
+                }
+                None
+            }
+            IdentityKind::Plookup => {
+                assert_eq!(
+                    identity.left.expressions.len(),
+                    identity.right.expressions.len()
+                );
+                identity.left.expressions.is_empty().then_some(index)
+            }
+            IdentityKind::Permutation => None,
+            IdentityKind::Connect => None,
+        })
+        .collect();
+    pil_file.remove_identities(&to_remove);
+}
+
+#[cfg(test)]
+mod test {
+    use number::GoldilocksField;
+    use pil_analyzer::pil_analyzer::process_pil_file_contents;
+
+    use crate::optimize;
+
+    use pretty_assertions::assert_eq;
+
+    #[test]
+    fn replace_fixed() {
+        let input = r#"namespace N(65536);
+    col fixed one = [1]*;
+    col fixed zero = [0]*;
+    col witness X;
+    col witness Y;
+    X * one = X * zero - zero + Y;
+    one * Y = zero * Y + 7 * X;
+"#;
+        let expectation = r#"namespace N(65536);
+    col witness X;
+    col witness Y;
+    N.X = N.Y;
+    N.Y = (7 * N.X);
+"#;
+        let optimized = optimize(process_pil_file_contents::<GoldilocksField>(input)).to_string();
```

Comment:
Maybe extract this into a function, make it generic over the field, and test with both Goldilocks and BN?

Reply:
BN is so expensive...